lib/Transforms/TupleToStruct.cpp

```diff
@@ -99,27 +99,14 @@ void TupleToStructPass::runOnOperation() {
     RewritePatternSet patterns(&context);
     ConversionTarget target(context);
 
-    target.addDynamicallyLegalOp<P4HIR::ReturnOp>(
-        [&](P4HIR::ReturnOp ret) { return typeConverter.isLegal(ret->getOperandTypes()); });
+    populateTypeConversionPattern(patterns, typeConverter);
 
-    target.markUnknownOpDynamicallyLegal([&](Operation *op) {
-        if (auto func = dyn_cast<FunctionOpInterface>(op)) {
-            auto fnType = func.getFunctionType();
-            return typeConverter.isLegal(fnType);
-        }
-
-        return typeConverter.isLegal(op->getOperandTypes()) &&
-               typeConverter.isLegal(op->getResultTypes());
-    });
+    patterns.add<TupleOpConversion, TupleExtractOpConversion>(typeConverter, &context);
```

Comment:
While now `configureUnknownOpDynamicallyLegalByTypes` handles the things via result type, I'd just explicitly do
```c++
    addIllegalOp<P4HIR::TupleOp, P4HIR::TupleExtractOp>();
```
just to clearly outline what is expected output :)